src/GraphQL/Utilities/Visitors/RemoveImportedTypesVisitor.cs

```diff
@@ -0,0 +1,113 @@
+using GraphQL.Types;
+using GraphQLParser.AST;
+using GraphQLParser.Visitors;
+
+namespace GraphQL.Utilities.Visitors;
+
+/// <summary>
+/// Remove all type and directive definitions that are imported from another schema.
+/// </summary>
+public sealed class RemoveImportedTypesVisitor : ASTVisitor<RemoveImportedTypesVisitor.Context>
+{
+    private static readonly RemoveImportedTypesVisitor _instance = new();
+
+    private RemoveImportedTypesVisitor()
+    {
+    }
+
+    /// <inheritdoc cref="RemoveFederationTypesVisitor"/>
+    public static void Visit(ASTNode node, ISchema schema)
+    {
+        var appliedDirectives = schema.GetAppliedDirectives();
+        if (appliedDirectives == null)
+            return;
+
+        List<string>? importedNamespaces = null;
+        HashSet<string>? importedTypes = null;
+        foreach (var appliedDirective in appliedDirectives)
+        {
+            var link = LinkConfiguration.GetConfiguration(appliedDirective);
+            if (link == null)
+                continue;
+            if (link.Namespace != null)
+            {
+                importedNamespaces ??= new();
+                importedNamespaces.Add(link.Namespace + "__");
+            }
+            if (link.Imports?.Count > 0)
+            {
+                importedTypes ??= new();
+                foreach (var import in link.Imports)
+                {
+                    importedTypes.Add(import.Value);
+                }
+            }
+        }
+
+        var context = new Context
+        {
+            ImportedNamespaces = importedNamespaces,
+            ImportedTypes = importedTypes,
+        };
+        _instance.VisitAsync(node, context).GetAwaiter().GetResult();
```

Comment:
![image](https://github.com/graphql-dotnet/graphql-dotnet/assets/4518483/081f9bcd-1aff-43ea-918a-07da75c156d6)

Reply:
`RemoveImportedTypesVisitor` is a sealed class and none of its methods call truly asynchronous methods.  In other words, it will always run synchronously.  Use of `.GetAwaiter().GetResult()` will unwrap any exceptions if they were to occur.